Swap the analyze and detect phases (#412)
buildpacks/docs: content/docs/reference/spec/migration/platform-api-0.6-0.7.md

+++
title="Platform API 0.6 -> 0.7"
+++

<!--more-->

This guide is most relevant to platform operators.

See the [spec release](https://github.com/buildpacks/spec/releases/tag/platform%2Fv0.7) for platform API 0.7 for the full list of changes and further details.

## Platform Operator

### Run the analyzer before the detector

> The order of the lifecycle phases has been changed starting from Platfrom API 0.7.\

Reviewer: ```suggestion
The order of the lifecycle phases has been changed starting from Platform API 0.7.\
```

Author: Please see https://github.com/buildpacks/docs/pull/412/commits/f343211c5ae42882b02dfcb04f197b716917e9d4.

The order before Platform API 0.7 was: detect, analyze, restore, build, and export.\
The order starting from Platform API 0.7 is: analyze, detect, restore, build, and export.\

> The main reason for this change was to validate registry access for all images that are used prior to the `detect` and `build` phases, and therefore provide faster failures for end users.

Reviewer: ```suggestion
The main reason for this change was to validate registry access for all images that are used during the build, and therefore provide faster failures for end users.
```

Author: Same comment as https://github.com/buildpacks/docs/pull/412#discussion_r706395550.

Author: What do you think about:

The main reason for this change was to validate registry access for all images that are used during the build as early as possible, instead of during the `build` phase. Starting from Platform API 0.7, it happens in the first lifecycle phase, as part of the `analyzer`, before running both the `detector` and the `builder`, and in this way it provides faster failures for end users.

Reviewer: > instead of during the `build` phase

I think in the past most of the failures would have happened in analyze and/or export, because those are the phases that need to talk to the registry. But otherwise this looks good.

Author: Please see https://github.com/buildpacks/docs/pull/412/commits/74d6a4cee786e2ee6473f37d70c8711759baeacb and let me know what you think.

> As part of this change, few flags were removed and others were added to some of the lifecycle phases.

Reviewer: ```suggestion
As part of this change, a few flags were removed and others were added to some of the lifecycle phases.
```

Author: Please see https://github.com/buildpacks/docs/pull/412/commits/f343211c5ae42882b02dfcb04f197b716917e9d4.

> The main reason for this change was to validate registry access for all images that are used during the build as early as possible. Starting from `Platform API 0.7`, it happens in the first lifecycle phase, as part of the `analyzer`, before running both the `detector` and the `builder`. In this way it provides faster failures for end users.

Reviewer: ```suggestion
The main reason for this change was to validate registry access for all images that are used during the build as early as possible. Starting from `Platform API 0.7`, it happens in the first lifecycle phase, as part of the `analyzer`, before running the `detector`. In this way it provides faster failures for end users.
```

Author: Same comment as https://github.com/buildpacks/docs/pull/412#discussion_r706438591.

Author: Please see https://github.com/buildpacks/docs/pull/412/commits/3fae08256362ad2fc1293dfba4a5543d1272e2fa.

The flags that were removed from the analyzer are:
* `-cache-dir`
* `-group`
* `-skip-layers`

The flags that were added to the analyzer are:
* `-previous-image`
* `-run-image`
* `-stack`
* `-tag`

The flags that were added to the restorer are:
* `-analyzed`
* `-skip-layers`

The flags that were removed from the exporter are:
* `-run-image`

> * `-stack`

Reviewer: ```suggestion
```
This is actually an error in the spec. We talked about removing stack, but then we realized that it's still needed.

Author: ohhhh...
I'll fix the document and will submit a PR to the spec repo.
Thanks!

Author: Please see https://github.com/buildpacks/docs/pull/412/commits/fd0fd6a537c5bfb0e8abe38a986d5bbc9aafa4a8 (the docs fix) and https://github.com/buildpacks/spec/pull/250 (spec PR).
